// @backstage/plugin-catalog/dist/index.esm.js
export { a as AboutContent, b as AboutField, d as CatalogEntityPage, e as CatalogIndexPage, o as CatalogKindHeader, p as CatalogSearchResultListItem, C as CatalogTable, D as DefaultStarredEntitiesApi, E as EntityAboutCard, g as EntityDependencyOfComponentsCard, h as EntityDependsOnComponentsCard, i as EntityDependsOnResourcesCard, j as EntityHasComponentsCard, k as EntityHasResourcesCard, l as EntityHasSubcomponentsCard, m as EntityHasSystemsCard, q as EntityLayout, n as EntityLinksCard, B as EntityListContainer, r as EntityOrphanWarning, t as EntityProcessingErrorsPanel, v as EntitySwitch, z as FilterContainer, F as FilteredEntityLayout, R as RelatedEntitiesCard, f as catalogPlugin, u as hasCatalogProcessingErrors, y as isComponentType, w as isKind, x as isNamespace, s as isOrphan } from './esm/index-24abea03.esm.js';
import 'zen-observable';
import '@backstage/catalog-model';
import 'lodash';
import '@backstage/core-components';
import '@backstage/core-plugin-api';
import '@backstage/integration-react';
import '@backstage/plugin-catalog-react';
import '@material-ui/core';
import '@material-ui/icons/Cached';
import '@material-ui/icons/Description';
import '@material-ui/icons/Edit';
import 'react';
import '@material-ui/icons/Language';
import 'pluralize';
import 'react-use/lib/useAsync';
import '@backstage/plugin-search-react';
import '@material-ui/core/styles';
import '@material-ui/icons/OpenInNew';
import '@material-ui/icons/Star';
import '@material-ui/icons/StarBorder';
import '@material-ui/lab';
import 'react-router-dom';
import '@material-ui/icons/BugReport';
import '@material-ui/icons/MoreVert';
import '@material-ui/icons/FileCopyTwoTone';
import '@backstage/plugin-catalog-common';
import '@material-ui/icons/Cancel';
import '@backstage/errors';
import '@backstage/catalog-client';
