// @rjsf/material-ui-v5/dist/material-ui.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@rjsf/core');
var React = require('react');
var AddIcon = require('@material-ui/icons/Add');
var IconButton = require('@material-ui/core/IconButton');
var Box = require('@material-ui/core/Box');
var Grid = require('@material-ui/core/Grid');
var Paper = require('@material-ui/core/Paper');
var utils = require('@rjsf/utils');
var TextField = require('@material-ui/core/TextField');
var Typography = require('@material-ui/core/Typography');
var ErrorIcon = require('@material-ui/icons/Error');
var List = require('@material-ui/core/List');
var ListItem = require('@material-ui/core/ListItem');
var ListItemIcon = require('@material-ui/core/ListItemIcon');
var ListItemText = require('@material-ui/core/ListItemText');
var ArrowUpwardIcon = require('@material-ui/icons/ArrowUpward');
var ArrowDownwardIcon = require('@material-ui/icons/ArrowDownward');
var RemoveIcon = require('@material-ui/icons/Remove');
var FormHelperText = require('@material-ui/core/FormHelperText');
var FormControl = require('@material-ui/core/FormControl');
var Button = require('@material-ui/core/Button');
var Divider = require('@material-ui/core/Divider');
var InputLabel = require('@material-ui/core/InputLabel');
var Input = require('@material-ui/core/Input');
var Checkbox = require('@material-ui/core/Checkbox');
var FormControlLabel = require('@material-ui/core/FormControlLabel');
var FormGroup = require('@material-ui/core/FormGroup');
var FormLabel = require('@material-ui/core/FormLabel');
var Radio = require('@material-ui/core/Radio');
var RadioGroup = require('@material-ui/core/RadioGroup');
var Slider = require('@material-ui/core/Slider');
var MenuItem = require('@material-ui/core/MenuItem');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var AddIcon__default = /*#__PURE__*/_interopDefaultLegacy(AddIcon);
var IconButton__default = /*#__PURE__*/_interopDefaultLegacy(IconButton);
var Box__default = /*#__PURE__*/_interopDefaultLegacy(Box);
var Grid__default = /*#__PURE__*/_interopDefaultLegacy(Grid);
var Paper__default = /*#__PURE__*/_interopDefaultLegacy(Paper);
var TextField__default = /*#__PURE__*/_interopDefaultLegacy(TextField);
var Typography__default = /*#__PURE__*/_interopDefaultLegacy(Typography);
var ErrorIcon__default = /*#__PURE__*/_interopDefaultLegacy(ErrorIcon);
var List__default = /*#__PURE__*/_interopDefaultLegacy(List);
var ListItem__default = /*#__PURE__*/_interopDefaultLegacy(ListItem);
var ListItemIcon__default = /*#__PURE__*/_interopDefaultLegacy(ListItemIcon);
var ListItemText__default = /*#__PURE__*/_interopDefaultLegacy(ListItemText);
var ArrowUpwardIcon__default = /*#__PURE__*/_interopDefaultLegacy(ArrowUpwardIcon);
var ArrowDownwardIcon__default = /*#__PURE__*/_interopDefaultLegacy(ArrowDownwardIcon);
var RemoveIcon__default = /*#__PURE__*/_interopDefaultLegacy(RemoveIcon);
var FormHelperText__default = /*#__PURE__*/_interopDefaultLegacy(FormHelperText);
var FormControl__default = /*#__PURE__*/_interopDefaultLegacy(FormControl);
var Button__default = /*#__PURE__*/_interopDefaultLegacy(Button);
var Divider__default = /*#__PURE__*/_interopDefaultLegacy(Divider);
var InputLabel__default = /*#__PURE__*/_interopDefaultLegacy(InputLabel);
var Input__default = /*#__PURE__*/_interopDefaultLegacy(Input);
var Checkbox__default = /*#__PURE__*/_interopDefaultLegacy(Checkbox);
var FormControlLabel__default = /*#__PURE__*/_interopDefaultLegacy(FormControlLabel);
var FormGroup__default = /*#__PURE__*/_interopDefaultLegacy(FormGroup);
var FormLabel__default = /*#__PURE__*/_interopDefaultLegacy(FormLabel);
var Radio__default = /*#__PURE__*/_interopDefaultLegacy(Radio);
var RadioGroup__default = /*#__PURE__*/_interopDefaultLegacy(RadioGroup);
var Slider__default = /*#__PURE__*/_interopDefaultLegacy(Slider);
var MenuItem__default = /*#__PURE__*/_interopDefaultLegacy(MenuItem);

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
    return target;
  };
  return _extends.apply(this, arguments);
}
function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;
  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }
  return target;
}

var _excluded$3 = ["uiSchema"];
var AddButton = function AddButton(_ref) {
  var props = _objectWithoutPropertiesLoose(_ref, _excluded$3);
  return /*#__PURE__*/React__default["default"].createElement(IconButton__default["default"], _extends({
    title: "Add Item"
  }, props, {
    color: "primary"
  }), /*#__PURE__*/React__default["default"].createElement(AddIcon__default["default"], null));
};

var ArrayFieldItemTemplate = function ArrayFieldItemTemplate(props) {
  var children = props.children,
    disabled = props.disabled,
    hasToolbar = props.hasToolbar,
    hasMoveDown = props.hasMoveDown,
    hasMoveUp = props.hasMoveUp,
    hasRemove = props.hasRemove,
    index = props.index,
    onDropIndexClick = props.onDropIndexClick,
    onReorderClick = props.onReorderClick,
    readonly = props.readonly,
    uiSchema = props.uiSchema,
    registry = props.registry;
  var _registry$templates$B = registry.templates.ButtonTemplates,
    MoveDownButton = _registry$templates$B.MoveDownButton,
    MoveUpButton = _registry$templates$B.MoveUpButton,
    RemoveButton = _registry$templates$B.RemoveButton;
  var btnStyle = {
    flex: 1,
    paddingLeft: 6,
    paddingRight: 6,
    fontWeight: "bold",
    minWidth: 0
  };
  return /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    alignItems: "center"
  }, /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true,
    xs: true,
    style: {
      overflow: "auto"
    }
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    mb: 2
  }, /*#__PURE__*/React__default["default"].createElement(Paper__default["default"], {
    elevation: 2
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    p: 2
  }, children)))), hasToolbar && /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true
  }, (hasMoveUp || hasMoveDown) && /*#__PURE__*/React__default["default"].createElement(MoveUpButton, {
    style: btnStyle,
    disabled: disabled || readonly || !hasMoveUp,
    onClick: onReorderClick(index, index - 1),
    uiSchema: uiSchema
  }), (hasMoveUp || hasMoveDown) && /*#__PURE__*/React__default["default"].createElement(MoveDownButton, {
    style: btnStyle,
    disabled: disabled || readonly || !hasMoveDown,
    onClick: onReorderClick(index, index + 1),
    uiSchema: uiSchema
  }), hasRemove && /*#__PURE__*/React__default["default"].createElement(RemoveButton, {
    style: btnStyle,
    disabled: disabled || readonly,
    onClick: onDropIndexClick(index),
    uiSchema: uiSchema
  })));
};

var _excluded$2 = ["key"];
var ArrayFieldTemplate = function ArrayFieldTemplate(props) {
  var canAdd = props.canAdd,
    disabled = props.disabled,
    idSchema = props.idSchema,
    uiSchema = props.uiSchema,
    items = props.items,
    onAddClick = props.onAddClick,
    readonly = props.readonly,
    registry = props.registry,
    required = props.required,
    schema = props.schema,
    title = props.title;
  var uiOptions = utils.getUiOptions(uiSchema);
  var ArrayFieldDescriptionTemplate = utils.getTemplate("ArrayFieldDescriptionTemplate", registry, uiOptions);
  var ArrayFieldItemTemplate = utils.getTemplate("ArrayFieldItemTemplate", registry, uiOptions);
  var ArrayFieldTitleTemplate = utils.getTemplate("ArrayFieldTitleTemplate", registry, uiOptions);
  // Button templates are not overridden in the uiSchema
  var AddButton = registry.templates.ButtonTemplates.AddButton;
  return /*#__PURE__*/React__default["default"].createElement(Paper__default["default"], {
    elevation: 2
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    p: 2
  }, /*#__PURE__*/React__default["default"].createElement(ArrayFieldTitleTemplate, {
    idSchema: idSchema,
    title: uiOptions.title || title,
    schema: schema,
    uiSchema: uiSchema,
    required: required,
    registry: registry
  }), /*#__PURE__*/React__default["default"].createElement(ArrayFieldDescriptionTemplate, {
    idSchema: idSchema,
    description: uiOptions.description || schema.description,
    schema: schema,
    uiSchema: uiSchema,
    registry: registry
  }), /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    key: "array-item-list-" + idSchema.$id
  }, items && items.map(function (_ref) {
    var key = _ref.key,
      itemProps = _objectWithoutPropertiesLoose(_ref, _excluded$2);
    return /*#__PURE__*/React__default["default"].createElement(ArrayFieldItemTemplate, _extends({
      key: key
    }, itemProps));
  }), canAdd && /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    justifyContent: "flex-end"
  }, /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    mt: 2
  }, /*#__PURE__*/React__default["default"].createElement(AddButton, {
    className: "array-item-add",
    onClick: onAddClick,
    disabled: disabled || readonly,
    uiSchema: uiSchema
  })))))));
};

var _excluded$1 = ["id", "placeholder", "required", "readonly", "disabled", "type", "label", "value", "onChange", "onBlur", "onFocus", "autofocus", "options", "schema", "uiSchema", "rawErrors", "formContext", "registry"],
  _excluded2$1 = ["step", "min", "max"];
var BaseInputTemplate = function BaseInputTemplate(_ref) {
  var id = _ref.id,
    placeholder = _ref.placeholder,
    required = _ref.required,
    readonly = _ref.readonly,
    disabled = _ref.disabled,
    type = _ref.type,
    label = _ref.label,
    value = _ref.value,
    onChange = _ref.onChange,
    onBlur = _ref.onBlur,
    onFocus = _ref.onFocus,
    autofocus = _ref.autofocus,
    options = _ref.options,
    schema = _ref.schema,
    uiSchema = _ref.uiSchema,
    _ref$rawErrors = _ref.rawErrors,
    rawErrors = _ref$rawErrors === void 0 ? [] : _ref$rawErrors,
    registry = _ref.registry,
    textFieldProps = _objectWithoutPropertiesLoose(_ref, _excluded$1);
  var inputProps = utils.getInputProps(schema, type, options);
  // Now we need to pull out the step, min, max into an inner `inputProps` for material-ui
  var step = inputProps.step,
    min = inputProps.min,
    max = inputProps.max,
    rest = _objectWithoutPropertiesLoose(inputProps, _excluded2$1);
  var otherProps = _extends({
    inputProps: _extends({
      step: step,
      min: min,
      max: max
    }, schema.examples ? {
      list: "examples_" + id
    } : undefined)
  }, rest);
  var _onChange = function _onChange(_ref2) {
    var value = _ref2.target.value;
    return onChange(value === "" ? options.emptyValue : value);
  };
  var _onBlur = function _onBlur(_ref3) {
    var value = _ref3.target.value;
    return onBlur(id, value);
  };
  var _onFocus = function _onFocus(_ref4) {
    var value = _ref4.target.value;
    return onFocus(id, value);
  };
  var schemaUtils = registry.schemaUtils;
  var displayLabel = schemaUtils.getDisplayLabel(schema, uiSchema);
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(TextField__default["default"], _extends({
    id: id,
    name: id,
    placeholder: placeholder,
    label: displayLabel ? label || schema.title : false,
    autoFocus: autofocus,
    required: required,
    disabled: disabled || readonly
  }, otherProps, {
    value: value || value === 0 ? value : "",
    error: rawErrors.length > 0,
    onChange: _onChange,
    onBlur: _onBlur,
    onFocus: _onFocus
  }, textFieldProps)), schema.examples && /*#__PURE__*/React__default["default"].createElement("datalist", {
    id: "examples_" + id
  }, schema.examples.concat(schema["default"] ? [schema["default"]] : []).map(function (example) {
    return /*#__PURE__*/React__default["default"].createElement("option", {
      key: example,
      value: example
    });
  })));
};

var DescriptionField = function DescriptionField(_ref) {
  var id = _ref.id,
    description = _ref.description;
  if (description) {
    return /*#__PURE__*/React__default["default"].createElement(Typography__default["default"], {
      id: id,
      variant: "subtitle2",
      style: {
        marginTop: "5px"
      }
    }, description);
  }
  return null;
};

var ErrorList = function ErrorList(_ref) {
  var errors = _ref.errors;
  return /*#__PURE__*/React__default["default"].createElement(Paper__default["default"], {
    elevation: 2
  }, /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    mb: 2,
    p: 2
  }, /*#__PURE__*/React__default["default"].createElement(Typography__default["default"], {
    variant: "h6"
  }, "Errors"), /*#__PURE__*/React__default["default"].createElement(List__default["default"], {
    dense: true
  }, errors.map(function (error, i) {
    return /*#__PURE__*/React__default["default"].createElement(ListItem__default["default"], {
      key: i
    }, /*#__PURE__*/React__default["default"].createElement(ListItemIcon__default["default"], null, /*#__PURE__*/React__default["default"].createElement(ErrorIcon__default["default"], {
      color: "error"
    })), /*#__PURE__*/React__default["default"].createElement(ListItemText__default["default"], {
      primary: error.stack
    }));
  }))));
};

var _excluded = ["icon", "color", "uiSchema"],
  _excluded2 = ["iconType"];
function MuiIconButton(props) {
  var icon = props.icon,
    color = props.color,
    otherProps = _objectWithoutPropertiesLoose(props, _excluded);
  return /*#__PURE__*/React__default["default"].createElement(IconButton__default["default"], _extends({}, otherProps, {
    size: "small",
    color: color
  }), icon);
}
function MoveDownButton(props) {
  return /*#__PURE__*/React__default["default"].createElement(MuiIconButton, _extends({
    title: "Move down"
  }, props, {
    icon: /*#__PURE__*/React__default["default"].createElement(ArrowDownwardIcon__default["default"], {
      fontSize: "small"
    })
  }));
}
function MoveUpButton(props) {
  return /*#__PURE__*/React__default["default"].createElement(MuiIconButton, _extends({
    title: "Move up"
  }, props, {
    icon: /*#__PURE__*/React__default["default"].createElement(ArrowUpwardIcon__default["default"], {
      fontSize: "small"
    })
  }));
}
function RemoveButton(props) {
  var iconType = props.iconType,
    otherProps = _objectWithoutPropertiesLoose(props, _excluded2);
  return /*#__PURE__*/React__default["default"].createElement(MuiIconButton, _extends({
    title: "Remove"
  }, otherProps, {
    color: "secondary",
    icon: /*#__PURE__*/React__default["default"].createElement(RemoveIcon__default["default"], {
      fontSize: iconType === "default" ? "medium" : "small"
    })
  }));
}

/** The `FieldErrorTemplate` component renders the errors local to the particular field
 *
 * @param props - The `FieldErrorProps` for the errors being rendered
 */
function FieldErrorTemplate(props) {
  var _props$errors = props.errors,
    errors = _props$errors === void 0 ? [] : _props$errors,
    idSchema = props.idSchema;
  if (errors.length === 0) {
    return null;
  }
  var id = idSchema.$id + "__error";
  return /*#__PURE__*/React__default["default"].createElement(List__default["default"], {
    dense: true,
    disablePadding: true
  }, errors.map(function (error, i) {
    return /*#__PURE__*/React__default["default"].createElement(ListItem__default["default"], {
      key: i,
      disableGutters: true
    }, /*#__PURE__*/React__default["default"].createElement(FormHelperText__default["default"], {
      id: id
    }, error));
  }));
}

/** The `FieldHelpTemplate` component renders any help desired for a field
 *
 * @param props - The `FieldHelpProps` to be rendered
 */
function FieldHelpTemplate(props) {
  var idSchema = props.idSchema,
    help = props.help;
  if (!help) {
    return null;
  }
  var id = idSchema.$id + "__help";
  return /*#__PURE__*/React__default["default"].createElement(FormHelperText__default["default"], {
    id: id
  }, help);
}

var FieldTemplate = function FieldTemplate(_ref) {
  var id = _ref.id,
    children = _ref.children,
    classNames = _ref.classNames,
    disabled = _ref.disabled,
    displayLabel = _ref.displayLabel,
    hidden = _ref.hidden,
    label = _ref.label,
    onDropPropertyClick = _ref.onDropPropertyClick,
    onKeyChange = _ref.onKeyChange,
    readonly = _ref.readonly,
    required = _ref.required,
    _ref$rawErrors = _ref.rawErrors,
    rawErrors = _ref$rawErrors === void 0 ? [] : _ref$rawErrors,
    errors = _ref.errors,
    help = _ref.help,
    rawDescription = _ref.rawDescription,
    schema = _ref.schema,
    uiSchema = _ref.uiSchema,
    registry = _ref.registry;
  var uiOptions = utils.getUiOptions(uiSchema);
  var WrapIfAdditionalTemplate = utils.getTemplate("WrapIfAdditionalTemplate", registry, uiOptions);
  if (hidden) {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      style: {
        display: "none"
      }
    }, children);
  }
  return /*#__PURE__*/React__default["default"].createElement(WrapIfAdditionalTemplate, {
    classNames: classNames,
    disabled: disabled,
    id: id,
    label: label,
    onDropPropertyClick: onDropPropertyClick,
    onKeyChange: onKeyChange,
    readonly: readonly,
    required: required,
    schema: schema,
    uiSchema: uiSchema,
    registry: registry
  }, /*#__PURE__*/React__default["default"].createElement(FormControl__default["default"], {
    fullWidth: true,
    error: rawErrors.length ? true : false,
    required: required
  }, children, displayLabel && rawDescription ? /*#__PURE__*/React__default["default"].createElement(Typography__default["default"], {
    variant: "caption",
    color: "textSecondary"
  }, rawDescription) : null, errors, help));
};

var ObjectFieldTemplate = function ObjectFieldTemplate(_ref) {
  var description = _ref.description,
    title = _ref.title,
    properties = _ref.properties,
    required = _ref.required,
    disabled = _ref.disabled,
    readonly = _ref.readonly,
    uiSchema = _ref.uiSchema,
    idSchema = _ref.idSchema,
    schema = _ref.schema,
    formData = _ref.formData,
    onAddClick = _ref.onAddClick,
    registry = _ref.registry;
  var uiOptions = utils.getUiOptions(uiSchema);
  var TitleFieldTemplate = utils.getTemplate("TitleFieldTemplate", registry, uiOptions);
  var DescriptionFieldTemplate = utils.getTemplate("DescriptionFieldTemplate", registry, uiOptions);
  // Button templates are not overridden in the uiSchema
  var AddButton = registry.templates.ButtonTemplates.AddButton;
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, (uiOptions.title || title) && /*#__PURE__*/React__default["default"].createElement(TitleFieldTemplate, {
    id: idSchema.$id + "-title",
    title: title,
    required: required,
    schema: schema,
    uiSchema: uiSchema,
    registry: registry
  }), (uiOptions.description || description) && /*#__PURE__*/React__default["default"].createElement(DescriptionFieldTemplate, {
    id: idSchema.$id + "-description",
    description: uiOptions.description || description,
    schema: schema,
    uiSchema: uiSchema,
    registry: registry
  }), /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    spacing: 2,
    style: {
      marginTop: "10px"
    }
  }, properties.map(function (element, index) {
    return (
      // Remove the <Grid> if the inner element is hidden as the <Grid>
      // itself would otherwise still take up space.
      element.hidden ? element.content : /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
        item: true,
        xs: 12,
        key: index,
        style: {
          marginBottom: "10px"
        }
      }, element.content)
    );
  }), utils.canExpand(schema, uiSchema, formData) && /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    justifyContent: "flex-end"
  }, /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true
  }, /*#__PURE__*/React__default["default"].createElement(AddButton, {
    className: "object-property-expand",
    onClick: onAddClick(schema),
    disabled: disabled || readonly,
    uiSchema: uiSchema
  })))));
};

var SubmitButton = function SubmitButton(props) {
  var _getSubmitButtonOptio = utils.getSubmitButtonOptions(props.uiSchema),
    submitText = _getSubmitButtonOptio.submitText,
    norender = _getSubmitButtonOptio.norender,
    submitButtonProps = _getSubmitButtonOptio.props;
  if (norender) {
    return null;
  }
  return /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    marginTop: 3
  }, /*#__PURE__*/React__default["default"].createElement(Button__default["default"], _extends({
    type: "submit",
    variant: "contained",
    color: "primary"
  }, submitButtonProps), submitText));
};

var TitleField = function TitleField(_ref) {
  var id = _ref.id,
    title = _ref.title;
  return /*#__PURE__*/React__default["default"].createElement(Box__default["default"], {
    id: id,
    mb: 1,
    mt: 1
  }, /*#__PURE__*/React__default["default"].createElement(Typography__default["default"], {
    variant: "h5"
  }, title), /*#__PURE__*/React__default["default"].createElement(Divider__default["default"], null));
};

var WrapIfAdditionalTemplate = function WrapIfAdditionalTemplate(_ref) {
  var children = _ref.children,
    classNames = _ref.classNames,
    disabled = _ref.disabled,
    id = _ref.id,
    label = _ref.label,
    onDropPropertyClick = _ref.onDropPropertyClick,
    onKeyChange = _ref.onKeyChange,
    readonly = _ref.readonly,
    required = _ref.required,
    schema = _ref.schema,
    uiSchema = _ref.uiSchema,
    registry = _ref.registry;
  // Button templates are not overridden in the uiSchema
  var RemoveButton = registry.templates.ButtonTemplates.RemoveButton;
  var keyLabel = label + " Key"; // i18n ?
  var additional = (utils.ADDITIONAL_PROPERTY_FLAG in schema);
  var btnStyle = {
    flex: 1,
    paddingLeft: 6,
    paddingRight: 6,
    fontWeight: "bold"
  };
  if (!additional) {
    return /*#__PURE__*/React__default["default"].createElement("div", {
      className: classNames
    }, children);
  }
  var handleBlur = function handleBlur(_ref2) {
    var target = _ref2.target;
    return onKeyChange(target.value);
  };
  return /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    container: true,
    key: id + "-key",
    alignItems: "center",
    spacing: 2,
    className: classNames
  }, /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true,
    xs: true
  }, /*#__PURE__*/React__default["default"].createElement(FormControl__default["default"], {
    fullWidth: true,
    required: required
  }, /*#__PURE__*/React__default["default"].createElement(InputLabel__default["default"], null, keyLabel), /*#__PURE__*/React__default["default"].createElement(Input__default["default"], {
    defaultValue: label,
    disabled: disabled || readonly,
    id: id + "-key",
    name: id + "-key",
    onBlur: !readonly ? handleBlur : undefined,
    type: "text"
  }))), /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true,
    xs: true
  }, children), /*#__PURE__*/React__default["default"].createElement(Grid__default["default"], {
    item: true
  }, /*#__PURE__*/React__default["default"].createElement(RemoveButton, {
    iconType: "default",
    style: btnStyle,
    disabled: disabled || readonly,
    onClick: onDropPropertyClick(label),
    uiSchema: uiSchema
  })));
};

var Templates = {
  ArrayFieldItemTemplate: ArrayFieldItemTemplate,
  ArrayFieldTemplate: ArrayFieldTemplate,
  BaseInputTemplate: BaseInputTemplate,
  ButtonTemplates: {
    AddButton: AddButton,
    MoveDownButton: MoveDownButton,
    MoveUpButton: MoveUpButton,
    RemoveButton: RemoveButton,
    SubmitButton: SubmitButton
  },
  DescriptionFieldTemplate: DescriptionField,
  ErrorListTemplate: ErrorList,
  FieldErrorTemplate: FieldErrorTemplate,
  FieldHelpTemplate: FieldHelpTemplate,
  FieldTemplate: FieldTemplate,
  ObjectFieldTemplate: ObjectFieldTemplate,
  TitleFieldTemplate: TitleField,
  WrapIfAdditionalTemplate: WrapIfAdditionalTemplate
};

var CheckboxWidget = function CheckboxWidget(props) {
  var schema = props.schema,
    id = props.id,
    value = props.value,
    disabled = props.disabled,
    readonly = props.readonly,
    label = props.label,
    autofocus = props.autofocus,
    onChange = props.onChange,
    onBlur = props.onBlur,
    onFocus = props.onFocus;
  // Because an unchecked checkbox will cause html5 validation to fail, only add
  // the "required" attribute if the field value must be "true", due to the
  // "const" or "enum" keywords
  var required = utils.schemaRequiresTrueValue(schema);
  var _onChange = function _onChange(_, checked) {
    return onChange(checked);
  };
  var _onBlur = function _onBlur(_ref) {
    var value = _ref.target.value;
    return onBlur(id, value);
  };
  var _onFocus = function _onFocus(_ref2) {
    var value = _ref2.target.value;
    return onFocus(id, value);
  };
  return /*#__PURE__*/React__default["default"].createElement(FormControlLabel__default["default"], {
    control: /*#__PURE__*/React__default["default"].createElement(Checkbox__default["default"], {
      id: id,
      name: id,
      checked: typeof value === "undefined" ? false : Boolean(value),
      required: required,
      disabled: disabled || readonly,
      autoFocus: autofocus,
      onChange: _onChange,
      onBlur: _onBlur,
      onFocus: _onFocus
    }),
    label: label || ""
  });
};

var selectValue = function selectValue(value, selected, all) {
  var at = all.indexOf(value);
  var updated = selected.slice(0, at).concat(value, selected.slice(at));
  // As inserting values at predefined index positions doesn't work with empty
  // arrays, we need to reorder the updated selection to match the initial order
  return updated.sort(function (a, b) {
    return all.indexOf(a) > all.indexOf(b);
  });
};
var deselectValue = function deselectValue(value, selected) {
  return selected.filter(function (v) {
    return v !== value;
  });
};
var CheckboxesWidget = function CheckboxesWidget(_ref) {
  var schema = _ref.schema,
    label = _ref.label,
    id = _ref.id,
    disabled = _ref.disabled,
    options = _ref.options,
    value = _ref.value,
    autofocus = _ref.autofocus,
    readonly = _ref.readonly,
    required = _ref.required,
    onChange = _ref.onChange,
    onBlur = _ref.onBlur,
    onFocus = _ref.onFocus;
  var enumOptions = options.enumOptions,
    enumDisabled = options.enumDisabled,
    inline = options.inline;
  var _onChange = function _onChange(option) {
    return function (_ref2) {
      var checked = _ref2.target.checked;
      var all = enumOptions.map(function (_ref3) {
        var value = _ref3.value;
        return value;
      });
      if (checked) {
        onChange(selectValue(option.value, value, all));
      } else {
        onChange(deselectValue(option.value, value));
      }
    };
  };
  var _onBlur = function _onBlur(_ref4) {
    var value = _ref4.target.value;
    return onBlur(id, value);
  };
  var _onFocus = function _onFocus(_ref5) {
    var value = _ref5.target.value;
    return onFocus(id, value);
  };
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(FormLabel__default["default"], {
    required: required,
    htmlFor: id
  }, label || schema.title), /*#__PURE__*/React__default["default"].createElement(FormGroup__default["default"], {
    id: id,
    row: !!inline
  }, Array.isArray(enumOptions) && enumOptions.map(function (option, index) {
    var checked = value.indexOf(option.value) !== -1;
    var itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
    var checkbox = /*#__PURE__*/React__default["default"].createElement(Checkbox__default["default"], {
      id: id + "-" + option.value,
      name: id,
      checked: checked,
      disabled: disabled || itemDisabled || readonly,
      autoFocus: autofocus && index === 0,
      onChange: _onChange(option),
      onBlur: _onBlur,
      onFocus: _onFocus
    });
    return /*#__PURE__*/React__default["default"].createElement(FormControlLabel__default["default"], {
      control: checkbox,
      key: option.value,
      label: option.label
    });
  })));
};

var DateWidget = function DateWidget(props) {
  var options = props.options,
    registry = props.registry;
  var BaseInputTemplate = utils.getTemplate("BaseInputTemplate", registry, options);
  return /*#__PURE__*/React__default["default"].createElement(BaseInputTemplate, _extends({
    type: "date",
    InputLabelProps: {
      shrink: true
    }
  }, props));
};

var DateTimeWidget = function DateTimeWidget(props) {
  var options = props.options,
    registry = props.registry;
  var BaseInputTemplate = utils.getTemplate("BaseInputTemplate", registry, options);
  var value = utils.utcToLocal(props.value);
  var onChange = function onChange(value) {
    props.onChange(utils.localToUTC(value));
  };
  return /*#__PURE__*/React__default["default"].createElement(BaseInputTemplate, _extends({
    type: "datetime-local",
    InputLabelProps: {
      shrink: true
    }
  }, props, {
    value: value,
    onChange: onChange
  }));
};

var RadioWidget = function RadioWidget(_ref) {
  var id = _ref.id,
    schema = _ref.schema,
    options = _ref.options,
    value = _ref.value,
    required = _ref.required,
    disabled = _ref.disabled,
    readonly = _ref.readonly,
    label = _ref.label,
    onChange = _ref.onChange,
    onBlur = _ref.onBlur,
    onFocus = _ref.onFocus;
  var enumOptions = options.enumOptions,
    enumDisabled = options.enumDisabled;
  var _onChange = function _onChange(_, value) {
    return onChange(schema.type == "boolean" ? value !== "false" : value);
  };
  var _onBlur = function _onBlur(_ref2) {
    var value = _ref2.target.value;
    return onBlur(id, value);
  };
  var _onFocus = function _onFocus(_ref3) {
    var value = _ref3.target.value;
    return onFocus(id, value);
  };
  var row = options ? options.inline : false;
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(FormLabel__default["default"], {
    required: required,
    htmlFor: id
  }, label || schema.title), /*#__PURE__*/React__default["default"].createElement(RadioGroup__default["default"], {
    id: id,
    name: id,
    value: "" + value,
    row: row,
    onChange: _onChange,
    onBlur: _onBlur,
    onFocus: _onFocus
  }, Array.isArray(enumOptions) && enumOptions.map(function (option) {
    var itemDisabled = Array.isArray(enumDisabled) && enumDisabled.indexOf(option.value) !== -1;
    var radio = /*#__PURE__*/React__default["default"].createElement(FormControlLabel__default["default"], {
      control: /*#__PURE__*/React__default["default"].createElement(Radio__default["default"], {
        name: id,
        id: id + "-" + option.value,
        color: "primary"
      }),
      label: "" + option.label,
      value: "" + option.value,
      key: option.value,
      disabled: disabled || itemDisabled || readonly
    });
    return radio;
  })));
};

var RangeWidget = function RangeWidget(_ref) {
  var value = _ref.value,
    readonly = _ref.readonly,
    disabled = _ref.disabled,
    onBlur = _ref.onBlur,
    onFocus = _ref.onFocus,
    options = _ref.options,
    schema = _ref.schema,
    onChange = _ref.onChange,
    required = _ref.required,
    label = _ref.label,
    id = _ref.id;
  var sliderProps = _extends({
    value: value,
    label: label,
    id: id,
    name: id
  }, utils.rangeSpec(schema));
  var _onChange = function _onChange(_, value) {
    onChange(value ? value : options.emptyValue);
  };
  var _onBlur = function _onBlur(_ref2) {
    var value = _ref2.target.value;
    return onBlur(id, value);
  };
  var _onFocus = function _onFocus(_ref3) {
    var value = _ref3.target.value;
    return onFocus(id, value);
  };
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(FormLabel__default["default"], {
    required: required,
    htmlFor: id
  }, label || schema.title), /*#__PURE__*/React__default["default"].createElement(Slider__default["default"], _extends({
    disabled: disabled || readonly,
    onChange: _onChange,
    onBlur: _onBlur,
    onFocus: _onFocus,
    valueLabelDisplay: "auto"
  }, sliderProps)));
};

var SelectWidget = function SelectWidget(_ref) {
  var schema = _ref.schema,
    id = _ref.id,
    options = _ref.options,
    label = _ref.label,
    required = _ref.required,
    disabled = _ref.disabled,
    readonly = _ref.readonly,
    value = _ref.value,
    multiple = _ref.multiple,
    autofocus = _ref.autofocus,
    onChange = _ref.onChange,
    onBlur = _ref.onBlur,
    onFocus = _ref.onFocus,
    _ref$rawErrors = _ref.rawErrors,
    rawErrors = _ref$rawErrors === void 0 ? [] : _ref$rawErrors;
  var enumOptions = options.enumOptions,
    enumDisabled = options.enumDisabled;
  var emptyValue = multiple ? [] : "";
  var _onChange = function _onChange(_ref2) {
    var value = _ref2.target.value;
    return onChange(utils.processSelectValue(schema, value, options));
  };
  var _onBlur = function _onBlur(_ref3) {
    var value = _ref3.target.value;
    return onBlur(id, utils.processSelectValue(schema, value, options));
  };
  var _onFocus = function _onFocus(_ref4) {
    var value = _ref4.target.value;
    return onFocus(id, utils.processSelectValue(schema, value, options));
  };
  return /*#__PURE__*/React__default["default"].createElement(TextField__default["default"], {
    id: id,
    name: id,
    label: label || schema.title,
    select: true,
    value: typeof value === "undefined" ? emptyValue : value,
    required: required,
    disabled: disabled || readonly,
    autoFocus: autofocus,
    error: rawErrors.length > 0,
    onChange: _onChange,
    onBlur: _onBlur,
    onFocus: _onFocus,
    InputLabelProps: {
      shrink: true
    },
    SelectProps: {
      multiple: typeof multiple === "undefined" ? false : multiple
    }
  }, enumOptions.map(function (_ref5, i) {
    var value = _ref5.value,
      label = _ref5.label;
    var disabled = enumDisabled && enumDisabled.indexOf(value) != -1;
    return /*#__PURE__*/React__default["default"].createElement(MenuItem__default["default"], {
      key: i,
      value: value,
      disabled: disabled
    }, label);
  }));
};

var TextareaWidget = function TextareaWidget(props) {
  var options = props.options,
    registry = props.registry;
  var BaseInputTemplate = utils.getTemplate("BaseInputTemplate", registry, options);
  var rows = 5;
  if (typeof options.rows === "string" || typeof options.rows === "number") {
    rows = options.rows;
  }
  return /*#__PURE__*/React__default["default"].createElement(BaseInputTemplate, _extends({}, props, {
    multiline: true,
    rows: rows
  }));
};

var Widgets = {
  CheckboxWidget: CheckboxWidget,
  CheckboxesWidget: CheckboxesWidget,
  DateWidget: DateWidget,
  DateTimeWidget: DateTimeWidget,
  RadioWidget: RadioWidget,
  RangeWidget: RangeWidget,
  SelectWidget: SelectWidget,
  TextareaWidget: TextareaWidget
};

var Theme = {
  templates: Templates,
  widgets: Widgets
};

var MuiForm = /*#__PURE__*/core.withTheme(Theme);

exports.Form = MuiForm;
exports.Templates = Templates;
exports.Theme = Theme;
exports.Widgets = Widgets;
exports["default"] = MuiForm;
