// @backstage/plugin-scaffolder/dist/index.esm.js
export { M as EntityNamePickerFieldExtension, K as EntityPickerFieldExtension, _ as EntityPickerFieldSchema, N as EntityTagsPickerFieldExtension, a2 as EntityTagsPickerFieldSchema, X as NextScaffolderPage, Q as OwnedEntityPickerFieldExtension, a1 as OwnedEntityPickerFieldSchema, P as OwnerPickerFieldExtension, $ as OwnerPickerFieldSchema, U as RepoUrlPickerFieldExtension, a0 as RepoUrlPickerFieldSchema, D as ScaffolderClient, H as ScaffolderFieldExtensions, J as ScaffolderLayouts, V as ScaffolderPage, m as TaskPage, T as TemplateTypePicker, Y as createNextScaffolderFieldExtension, G as createScaffolderFieldExtension, I as createScaffolderLayout, Z as makeFieldSchemaFromZod, C as nextRouteRef, B as nextSelectedTemplateRouteRef, w as repoPickerValidation, f as rootRouteRef, c as scaffolderApiRef, W as scaffolderPlugin, s as selectedTemplateRouteRef, a3 as useTemplateSecrets } from './esm/index-8dc22fc0.esm.js';
import '@backstage/catalog-model';
import '@backstage/core-plugin-api';
import '@backstage/errors';
import 'qs';
import 'zen-observable';
import '@backstage/plugin-catalog-react';
import '@material-ui/core';
import '@material-ui/core/FormControl';
import '@material-ui/lab/Autocomplete';
import 'react';
import 'react-use/lib/useAsync';
import 'zod';
import 'zod-to-json-schema';
import 'react-use/lib/useEffectOnce';
import '@material-ui/lab';
import '@backstage/integration-react';
import '@material-ui/core/FormHelperText';
import '@material-ui/core/Input';
import '@material-ui/core/InputLabel';
import '@backstage/core-components';
import 'react-use/lib/useDebounce';
import 'lodash/capitalize';
import '@material-ui/icons/CheckBox';
import '@material-ui/icons/CheckBoxOutlineBlank';
import '@material-ui/icons/ExpandMore';
import '@material-ui/core/Grid';
import '@material-ui/core/Step';
import '@material-ui/core/StepLabel';
import '@material-ui/core/Stepper';
import '@material-ui/core/styles';
import '@material-ui/core/Typography';
import '@material-ui/icons/Cancel';
import '@material-ui/icons/Check';
import '@material-ui/icons/FiberManualRecord';
import 'classnames';
import 'luxon';
import 'react-router-dom';
import 'react-use/lib/useInterval';
import 'use-immer';
import '@material-ui/icons/Language';
